modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/viewmodel/ShelfViewModel.kt

```diff
@@ -0,0 +1,206 @@
+package au.com.shiftyjelly.pocketcasts.player.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import au.com.shiftyjelly.pocketcasts.analytics.AnalyticsEvent
+import au.com.shiftyjelly.pocketcasts.analytics.AnalyticsTracker
+import au.com.shiftyjelly.pocketcasts.models.entity.BaseEpisode
+import au.com.shiftyjelly.pocketcasts.models.to.Transcript
+import au.com.shiftyjelly.pocketcasts.preferences.Settings
+import au.com.shiftyjelly.pocketcasts.preferences.model.ShelfItem
+import au.com.shiftyjelly.pocketcasts.preferences.model.ShelfRowItem
+import au.com.shiftyjelly.pocketcasts.preferences.model.ShelfTitle
+import au.com.shiftyjelly.pocketcasts.repositories.chromecast.ChromeCastAnalytics
+import au.com.shiftyjelly.pocketcasts.repositories.podcast.TranscriptsManager
+import au.com.shiftyjelly.pocketcasts.utils.featureflag.Feature
+import au.com.shiftyjelly.pocketcasts.utils.featureflag.FeatureFlag
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import dagger.hilt.android.lifecycle.HiltViewModel
+import java.util.Collections
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.distinctUntilChangedBy
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import au.com.shiftyjelly.pocketcasts.localization.R as LR
+
+@HiltViewModel(assistedFactory = ShelfViewModel.Factory::class)
+class ShelfViewModel @AssistedInject constructor(
+    @Assisted private val episodeId: String?,
```

Comment:
Is it correct for this view model to have a `null` `episodeId`? It seems like many data-handling issues arise from this. Would it be possible to address this at the very first entry point, perhaps in `ShelfBottomSheet.newInstance()`, and avoid creating the fragment if there’s no episode ID?

Reply:
I checked we can make it non-nullable. 
72dbf76